### ISSUE TITLE
Loss and accuracy fluctuating too much on my custom data.

### ISSUE THREAD

**Issue by marmikp** [dreamquark-ai/tabnet#230]
Wonderful work!!

I have some concerns regarding the training of tabnet on my own data.
I am trying to train tabnet on my custom data which has 43 categorical columns and 3 numerical columns, with 2+lac of sample splited in 80/20 training, and testing set. Parameters for training are {n_d=256, n_a=256, n_steps=5, gamma=1.5, n_independent=2, n_shared=2,lambda_sparse=1e-4, momentum=0.3,clip_value=2., optimizer_fn=torch.optim.Adam, optimizer_params=dict(lr=2e-2), max_epochs=1000, patience=100, batch_size=16384, virtual_batch_size=256}.

While training, accuracy, and loss are fluctuating very much. Example is here. 
epoch 229| loss: 0.00691 | train_accuracy: 0.17529 | valid_accuracy: 0.22991 |  10:45:38s
epoch 230| loss: 0.0096  | train_accuracy: 0.16748 | valid_accuracy: 0.32664 |  10:48:27s
epoch 231| loss: 0.00761 | train_accuracy: 0.60939 | valid_accuracy: 0.86203 |  10:51:15s
epoch 232| loss: 0.01144 | train_accuracy: 0.28538 | valid_accuracy: 0.37121 |  10:54:04s
epoch 233| loss: 0.0069  | train_accuracy: 0.4226  | valid_accuracy: 0.67471 |  10:56:53s
epoch 234| loss: 0.00589 | train_accuracy: 0.64774 | valid_accuracy: 0.76433 |  10:59:42s
epoch 235| loss: 0.00551 | train_accuracy: 0.83802 | valid_accuracy: 0.91635 |  11:02:31s

So I wanted to know that is this normal while training? and if yes then why?
Thank you

**Comment by eduardocarvp** [comment]
Hi, @marmikp . I noticed that in the beginning fluctuations can occur due mostly to the batch-normalization layers that have not yet achieved some kind of convergence/stability. But that does not seem to be your case, since we're talking about epoch 230+.

Could you give more details on the problem you're trying to solve, please? Multiclass classification? Train/valid shuffled before split? Have you tested on other datasets where this happens as well?

**Comment by Optimox** [comment]
Hello @marmikp 

How much data sample does your validation set has?
How many classes do you predict in your problem?
What's the class imbalance you have?

There are many reasons for an accuracy to fluctuate as it's very sensitive due to thresholding. Maybe you could also try to add a learning rate decay, which will stabilize your training after some epochs.

The loss itself seems to be fluctuating a bit. Maybe you reached a point where it's hard to distinguish between classes but it seems odd that the model won't learn perfectly on the training set. Are you sure about your labels?

**Comment by marmikp** [comment]
Thanks a lot for your quick reply.

Yes, I am training a model for the multiclass classification problem. There are 46 classes and I've shuffled all the data before splitting in train, test.
I've trained my custom neural network classifier with these data. And it reaches 94% accuracy without fluctuation in the accuracy.

2001046 samples are available after oversampling in training data. (167190 sample without oversampling) - Oversampling due to imbalanced data.
18577 samples in the validation set.
46442 samples in the testing set.

Yes, label data are okay.

**Comment by Optimox** [comment]
@marmikp any reason to set such a high momentum?

With that amount of data I think you shouldn't need such a large number of epochs.

What I would try :
- leave momentum as default
- Step LR of 0.9 per epoch (train for 20 - 50 epochs max)
- you could also try to set a larger batch size to make things faster : maybe 16384
- do not oversample by adding rows to train, start with raw data and set weights=1 during training
- use a GPU (not sure if you already are or if you can)

Let us know if things are getting better!

**Comment by marmikp** [comment]
@Optimox 

- We have used momentum as the default value used in forest_example.ipynb. 
- Does step_lr mean step_size in the scheduler parameter? It is 20 in our example.
- We've already used batch_size of 16384.
- You told me that I should set weights = 1, but I am confused about which weights? Are these model parameters?
- Yes, I am using GPU.

Thank you :)

**Comment by eduardocarvp** [comment]
As for the weights, it is a `fit` parameter:
```
weights : bool or dictionnary
    0 for no balancing
    1 for automated balancing
    dict for custom weights per class
```
When it is 1 it will use the inverse of the  occurrences of each modality on the target:

```
if isinstance(weights, int):
        if weights == 0:
            need_shuffle = True
            sampler = None
        elif weights == 1:
            need_shuffle = False
            class_sample_count = np.array(
                [len(np.where(y_train == t)[0]) for t in np.unique(y_train)]
            )

            weights = 1.0 / class_sample_count

            samples_weight = np.array([weights[t] for t in y_train])

            samples_weight = torch.from_numpy(samples_weight)
            samples_weight = samples_weight.double()
            sampler = WeightedRandomSampler(samples_weight, len(samples_weight))
        else:
            raise ValueError("Weights should be either 0, 1, dictionnary or list.")
```

**Comment by Optimox** [comment]
I think you can decay every epoch
with something like this :
```python
scheduler_params = {"gamma": 0.95,
                                   "step_size": 1}
scheduler_fn=torch.optim.lr_scheduler.StepLR
```

As for the balancing weights see @eduardocarvp answer above, it's a sampling parameter to train more often on rare classes and so it helps dealing with unbalanced data.

**Comment by marmikp** [comment]
Thank you very much for your support.
We have relatively lesser fluctuation with these parameters.

Thanks a lot :)
